documentation/design-docs/diagnostics-client-library.md

```diff
@@ -240,6 +240,110 @@ public static int AttachProfiler(int processId, Guid profilerGuid, string profil
 }
 ```
 
+#### 8. Write a simple monitor agent
+
+This sample shows how to use the `DiagnosticsAgent` API to create a simple Agent application.
+```cs
+namespace Microsoft.Diagnostics.Tools.Reverse
+{
+    class Program
+    {
+        static ConcurrentDictionary<Guid, (Task, EventPipeSession)> sessionDict = new ConcurrentDictionary<int, (Task, EventPipeSession)>();
+        static ConcurrentDictionary<Guid, (int, DiagnosticsClient)> clientDict = new ConcurrentDictionary<int, (int, DiagnosticsClient)>();
+
+        static void Main(string[] args)
+        {
+            // get the address for the server (this sample assumes Linux)
+            string address = (args.Length >= 1) ? args[0] : "~/myawesome.sock";
+            try
+            {
+                using (DiagnosticsAgent agent = new DiagnosticsAgent(address))
+                {
+                    agent.OnDiagnosticsConnection += (sender, eventArgs) =>
+                    {
+                        clientDict[eventArgs.RuntimeInstanceCookie] = (eventArgs.ProcessId, eventArgs.Client);
```

Comment:
Will there be some mechanism by which multiple DiagnosticsClients can be established to the same process from one DiagnosticsAgent? I thinking that dotnet-monitory might need this if, for example, there is a active session for capturing logs, metrics, etc and then there is another request to capture dumps. I don't know if dotnet-monitor would be able to use the same connection concurrently for different collections of data.

Reply:
I agree, I don't think a connection should include the DiagnosticClient. I'd suggest the callback arg/return value can be used to create an arbitrary number of DiagnosticClients. For example:
````
var agent = new DiagnosticsAgent(address);
var runtimeEndpoint = await agent.ListenForRuntimesAsync();
Console.WriteLine("Dotnet process detected with pid {runtimeEndpoint.Pid}");
// only listening to one for simplicity

while(true)
{
    DotnetMonitorRequest r = await webserver.GetNextRequestAsync();
    DiagnosticClient c = new DiagnosticClient(runtimeEndpoint);
    Task.Run(() => HandleRequest(r,c));
}

Reply:
I don't think there's anything preventing you from using the same `DiagnosticsClient` for multiple actions at once, even in the current implementation.  You should already be able to start N<64 tracing sessions per process at once, and use the same `DiagnosticsClient` to capture dumps, gcdumps, etc. at the same time.

Is the suggestion that `DiagnosticClient`s should be disposable, e.g., use once, constructs?  Right now, `DiagnosticClient`s abstract the connection process entirely for both traditional and reversed connections.  Every time you issue a command for a traditional connection, it reconnects to the underlying transport.  This new iteration, effectively does the same thing for reverse connections, but does so by caching the reverse connection each time the runtime advertises.

Reply:
> I don't think there's anything preventing you from using the same DiagnosticsClient for multiple actions at once, even in the current implementation

You could design it to work, but by convention .NET types are not considered thread-safe unless specifically documented otherwise. We could implement and document that DiagnosticsClient is thread-safe, but people will probably assume that it isn't by default. I'd recommend going with convention is easier than trying to explain that the API supports free-threaded use.

> Is the suggestion that DiagnosticClients should be disposable, e.g., use once, constructs?

I think we should allow, but not require, that kind of usage.